File: app/src/main/java/lt/bulevicius/tessonetapp/app/ApplicationModule.java
package lt.bulevicius.tessonetapp.app;

import android.content.Context;
import android.content.SharedPreferences;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import lt.bulevicius.tessonetapp.storage.LocalDataProvider;
import lt.bulevicius.tessonetapp.storage.SharedPreferenceProvider;
import timber.log.Timber;

@Module
@SuppressWarnings("all")
public final class ApplicationModule {

    private final Context context;

    public ApplicationModule(Context context) {
        this.context = context;
    }

    @Provides
    @Singleton
    public final Context context() {
        return context;
    }

    @Provides
    @Singleton
    public final Timber.Tree timber() {
        return new Timber.DebugTree();
    }

    @Provides
    @Singleton
    public final LocalDataProvider localDataProvider(SharedPreferences sharedPreferences) {
        return new SharedPreferenceProvider(sharedPreferences);
    }

    @Provides
    @Singleton
    public final SharedPreferences sharedPreferences(Context context) {
        return context.getSharedPreferences(
                "lt.bulevicius.tessonetapp", Context.MODE_PRIVATE);
    }
}

File: app/src/main/java/com/giedrius/androidparty/task/dependency/ApplicationModule.kt
package com.giedrius.androidparty.task.dependency

import android.content.Context
import android.content.SharedPreferences
import com.giedrius.androidparty.task.App
import com.giedrius.androidparty.task.server.login.LoginClient
import com.giedrius.androidparty.task.server.login.LoginClientImplementation
import com.giedrius.androidparty.task.data.Repository
import com.giedrius.androidparty.task.data.RepositoryImplementation
import com.giedrius.androidparty.task.utils.Constants
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

@Module
class ApplicationModule {
    @Provides fun provideRepository(loginClient: LoginClient, sharedPreferences: SharedPreferences): Repository = RepositoryImplementation.getInstance(loginClient, sharedPreferences)
    @Provides fun provideLoginClient(): LoginClient = LoginClientImplementation()
    @Provides @Singleton fun provideSharedPreferences(context: Context): SharedPreferences = context.getSharedPreferences(Constants.SHARED_PREFERENCES_NAME, Context.MODE_PRIVATE)
    @Provides @Singleton fun provideContext(app: App): Context = app.applicationContext
}